cmd/root.go

```diff
@@ -0,0 +1,142 @@
+var repoURL *url.URL
+var logger utils.Logger
+var verbose bool
+
+const rootDesc = `Mirror Helm Charts from an index file into a local folder.
+
+For example:
+
+helm-mirror https://yourorg.com/charts /yourorg/charts
+
+This will download the index file and the charts in to
+the folder indicated.
+
+The index file is a yaml that contains a list of
+charts in this format. Example:
+
+	apiVersion: v1
+	entries:
+	  chart:
+	  - apiVersion: 1.0.0
+	    created: 2018-08-08T00:00:00.00000000Z
+	    description: A Helm chart for your application
+	    digest: 3aa68d6cb66c14c1fcffc6dc6d0ad8a65b90b90c10f9f04125dc6fcaf8ef1b20
+	    name: chart
+	    urls:
+	    - https://kubernetes-charts.yourorganization.com/chart-1.0.0.tgz
+	  chart2:
+	  - apiVersion: 1.0.0
+	    created: 2018-08-08T00:00:00.00000000Z
+	    description: A Helm chart for your application
+	    digest: 7ae62d60b61c14c1fcffc6dc670e72e62b91b91c10f9f04125dc67cef2ef0b21
+	    name: chart
+	    urls:
+	    - https://kubernetes-charts.yourorganization.com/chart2-1.0.0.tgz
+
+This will download these charts
+
+	https://kubernetes-charts.yourorganization.com/chart-1.0.0.tgz
+	https://kubernetes-charts.yourorganization.com/chart2-1.0.0.tgz
+
+into your destination folder.`
+
+// rootCmd represents the base command when called without any subcommands
+var rootCmd = &cobra.Command{
+	Use:   "mirror [Repo URL] [Destination Folder]",
+	Short: "Mirror Helm Charts from an index file into a local folder.",
+	Long:  rootDesc,
+	Args:  validateRootArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		logger = utils.NewLogger(verbose, os.Stdout)
+		err := os.MkdirAll(folder, 0777)
+		if err != nil {
+			logger.Fatal("creating destination folder: %s", err)
+			return err
+		}
+		err = os.Chdir(folder)
+		if err != nil {
+			logger.Fatal("while moving to destination folder: %s", err)
+			return err
+		}
+		io := utils.NewIo(logger, 0)
+		web := utils.NewWeb(logger)
+
+		getService := service.NewGetService(repoURL.String(), logger, io, web)
+		err = getService.Get()
+		if err != nil {
+			return err
+		}
+		return nil
+	},
+}
+
+// Execute adds all child commands to the root command and sets flags appropriately.
+// This is called by main.main(). It only needs to happen once to the rootCmd.
+func Execute() {
+	if err := rootCmd.Execute(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+func init() {
+	rootCmd.PersistentFlags().BoolVarP(&verbose, "verbose", "v", false, "verbose output")
+	rootCmd.AddCommand(newVersionCmd())
+}
+
+func validateRootArgs(cmd *cobra.Command, args []string) error {
+	l := utils.NewLogger(false, os.Stdout)
+	if len(args) < 2 {
+		if len(args) == 1 && args[0] == "help" {
+			return nil
+		}
+		l.Fatal("requires at least two args to execute")
+		return errors.New("requires at least two args to execute")
+	}
+	url, err := url.Parse(args[0])
+	if err != nil {
+		l.Fatal("not a valid URL for index file: %s", err)
+		return err
+	}
+	if url.Scheme == "" {
```

Comment:
not sure how `url.Parse` handles this, but would it make sense to check for `if !url.Scheme == "https" || !url.Scheme == "http"` ?

Reply:
yes, when I wrote it I was just going for a generic, but in this case it makes sense to have it test just http and https